<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_13) on Tue Aug 04 17:00:19 CEST 2009 -->
<TITLE>
Data
</TITLE>

<META NAME="date" CONTENT="2009-08-04">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="Data";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/Data.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../data/Filter.html" title="class in data"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../index.html?data/Data.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Data.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
data</FONT>
<BR>
Class Data</H2>
<PRE>
java.lang.Object
  <IMG SRC="../resources/inherit.gif" ALT="extended by "><B>data.Data</B>
</PRE>
<HR>
<DL>
<DT><PRE>public class <B>Data</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
Reading and containing RNA values from an input file and providing methods for the calculation of RNA ratios 
 as well as methods to save the filtered values to a file
<P>

<P>
<DL>
<DT><B>Author:</B></DT>
  <DD>Stefanie Kaufmann</DD>
</DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#absent">absent</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#absentLabel">absentLabel</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#AVERAGEFIRST">AVERAGEFIRST</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#geneName">geneName</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#RATIOFIRST">RATIOFIRST</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#spotid">spotid</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../data/Data.html#Data(double[][], double[][], double[][], java.lang.String[], java.lang.String[], java.lang.String[], data.Mapping)">Data</A></B>(double[][]&nbsp;totalRNA,
     double[][]&nbsp;newRNA,
     double[][]&nbsp;preRNA,
     java.lang.String[]&nbsp;totalDescr,
     java.lang.String[]&nbsp;newDescr,
     java.lang.String[]&nbsp;preDescr,
     <A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;&nbsp;map)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a Data-Object based on already existing data</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../data/Data.html#Data(java.lang.String, java.util.List, java.util.List, java.util.List)">Data</A></B>(java.lang.String&nbsp;file,
     java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
     java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
     java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a Data-Object based on an Input-File and the list of columns to be read with no additional attributes</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../data/Data.html#Data(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List)">Data</A></B>(java.lang.String&nbsp;file,
     java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
     java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
     java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
     java.util.List&lt;java.lang.String&gt;&nbsp;attr)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a Data-Object based on an Input-File and the list of columns to be read, defines columns containing attributes</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#addAttributes(java.lang.String)">addAttributes</A></B>(java.lang.String&nbsp;inputfile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads in a file containing two columns: column 1: spotID, column2: attribute (e.g. gene name)
 The resulting Mapping is added to the List attrMaps and accessible over the index received from attrMapsDescr,
 where the name of the column is stored</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#calculateAverageFirst(double[][], double[][])">calculateAverageFirst</A></B>(double[][]&nbsp;rna,
                      double[][]&nbsp;total)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;calculates the average values from all replicates for the linear regression
 calculates the ratio of the averages (AVERAGEFIRST method)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#calculateNewMapping()">calculateNewMapping</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculates the Mapping corresponding to the ratio</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#calculateRatio(double[][], double[][])">calculateRatio</A></B>(double[][]&nbsp;rna,
               double[][]&nbsp;total)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculates filtered Ratio of a set of RNA values (e.g. newly transcribed RNA) to a second set of values (e.g. total RNA)
 based on the specified method to calculate the average of multiple replicates.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#calculateRatioFirst(double[][])">calculateRatioFirst</A></B>(double[][]&nbsp;allRatio)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;calculates the average values from all replicates for the linear regression
 calculates the average of the ratios (RATIOFIRST method)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#copyMap()">copyMap</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Copies the content of the Mapping object</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#evaluate(java.lang.String, int)">evaluate</A></B>(java.lang.String&nbsp;fasta,
         int&nbsp;column)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Method for the Uracil-based evaluation of the values</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getAttrDescr()">getAttrDescr</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the description of the attributes</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.ArrayList&lt;<A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.String&gt;&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getAttributes()">getAttributes</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the attribute lists</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getAvgLo()">getAvgLo</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the average log(e'/n')</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getAvgUr()">getAvgUr</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the average number of Uracil</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getFilename()">getFilename</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the name of the file containing the data</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getMap()">getMap</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the map corresponding to the data</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getNewDescr()">getNewDescr</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the description = column labels of newly transcribed RNA</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getNewRNA()">getNewRNA</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns all the values for newly transcribed RNA</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getNumberExperiments()">getNumberExperiments</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the number of measured experiments (not changeable through filtering!)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getPlotName()">getPlotName</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the name of the file into which evaluation results will be printed
 as a plotable table</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getPqs()">getPqs</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the probe set quality scores</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getPreDescr()">getPreDescr</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the description = column labels of pre-existing RNA</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getPreRNA()">getPreRNA</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns all the values for pre-existing RNA</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getTotalDescr()">getTotalDescr</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the description = column labels of total RNA</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#getTotalRNA()">getTotalRNA</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns all the values for total RNA</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#readInput(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List)">readInput</A></B>(java.lang.String&nbsp;file,
          java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
          java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
          java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
          java.util.List&lt;java.lang.String&gt;&nbsp;addit)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Takes an input file, reads the values and saves them in the class variables of the Data object</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#restrictRatio(double[][], int)">restrictRatio</A></B>(double[][]&nbsp;input,
              int&nbsp;replicate)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Restricts a multidimensional array onto an array with one dimension containing the wanted replicate</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#setAbsent(java.lang.String)">setAbsent</A></B>(java.lang.String&nbsp;absent)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the name of the column (column label) which holds information on present/absent calls</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#setAttrDescr(java.lang.String[])">setAttrDescr</A></B>(java.lang.String[]&nbsp;attrDescr)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the description of all attributes</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#setAttributes(java.util.ArrayList)">setAttributes</A></B>(java.util.ArrayList&lt;<A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.String&gt;&gt;&nbsp;attributes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the attributes list for the data</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#setFilename(java.lang.String)">setFilename</A></B>(java.lang.String&nbsp;filename)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the name of the file holding the data information</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#setGeneName(java.lang.String)">setGeneName</A></B>(java.lang.String&nbsp;geneName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the name of the column which holds information on the gene name</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#setMap(data.Mapping)">setMap</A></B>(<A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;&nbsp;map)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the mapping corresponding to the whole data and filtered values</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#setMethod(int)">setMethod</A></B>(int&nbsp;i)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the method used for calculating the average ratios over all replicates</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#setPqs(double[])">setPqs</A></B>(double[]&nbsp;pqs)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the probe set quality scores</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#setUnequal(boolean)">setUnequal</A></B>(boolean&nbsp;unequal)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Set the value of the unequal variable, meaning if there is an
 unequal number of measurements (new, pre, total)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#writeOutput(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List)">writeOutput</A></B>(java.lang.String&nbsp;output,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
            java.util.List&lt;java.lang.String&gt;&nbsp;addit)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;desired columns are written to a file; former order is not kept, instead all values of one type are added sequentially</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../data/Data.html#writeToFileArray(java.lang.String, double[])">writeToFileArray</A></B>(java.lang.String&nbsp;output,
                 double[]&nbsp;array)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes the half life values contained in the object and in the map together with corresponding spotIds into a file
 Appropriate for half lives, ratios & pqs (for which the new mapping applies)</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="RATIOFIRST"><!-- --></A><H3>
RATIOFIRST</H3>
<PRE>
public static final int <B>RATIOFIRST</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../constant-values.html#data.Data.RATIOFIRST">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="AVERAGEFIRST"><!-- --></A><H3>
AVERAGEFIRST</H3>
<PRE>
public static final int <B>AVERAGEFIRST</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../constant-values.html#data.Data.AVERAGEFIRST">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="absentLabel"><!-- --></A><H3>
absentLabel</H3>
<PRE>
public static java.lang.String <B>absentLabel</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="absent"><!-- --></A><H3>
absent</H3>
<PRE>
public static java.lang.String <B>absent</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="geneName"><!-- --></A><H3>
geneName</H3>
<PRE>
public static java.lang.String <B>geneName</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="spotid"><!-- --></A><H3>
spotid</H3>
<PRE>
public static java.lang.String <B>spotid</B></PRE>
<DL>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="Data(java.lang.String, java.util.List, java.util.List, java.util.List)"><!-- --></A><H3>
Data</H3>
<PRE>
public <B>Data</B>(java.lang.String&nbsp;file,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre)</PRE>
<DL>
<DD>Constructs a Data-Object based on an Input-File and the list of columns to be read with no additional attributes
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>file</CODE> - Name of the Input-file<DD><CODE>columnsTotal</CODE> - List containing the labels of columns for total RNA<DD><CODE>columnsNew</CODE> - List containing the labels of columns for newly transcribed RNA<DD><CODE>columnsPre</CODE> - List containing the labels of columns for pre-existing RNA</DL>
</DL>
<HR>

<A NAME="Data(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List)"><!-- --></A><H3>
Data</H3>
<PRE>
public <B>Data</B>(java.lang.String&nbsp;file,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
            java.util.List&lt;java.lang.String&gt;&nbsp;attr)</PRE>
<DL>
<DD>Constructs a Data-Object based on an Input-File and the list of columns to be read, defines columns containing attributes
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>file</CODE> - Name of the Input-file<DD><CODE>columnsTotal</CODE> - List containing the labels of columns for total RNA<DD><CODE>columnsNew</CODE> - List containing the labels of columns for newly transcribed RNA<DD><CODE>columnsPre</CODE> - List containing the labels of columns for pre-existing RNA<DD><CODE>attr</CODE> - List containing the labels of columns for attributes</DL>
</DL>
<HR>

<A NAME="Data(double[][], double[][], double[][], java.lang.String[], java.lang.String[], java.lang.String[], data.Mapping)"><!-- --></A><H3>
Data</H3>
<PRE>
public <B>Data</B>(double[][]&nbsp;totalRNA,
            double[][]&nbsp;newRNA,
            double[][]&nbsp;preRNA,
            java.lang.String[]&nbsp;totalDescr,
            java.lang.String[]&nbsp;newDescr,
            java.lang.String[]&nbsp;preDescr,
            <A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;&nbsp;map)</PRE>
<DL>
<DD>Constructs a Data-Object based on already existing data
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>totalRNA</CODE> - Array containing all values of total RNA<DD><CODE>newRNA</CODE> - Array containing all values of newly transcribed RNA<DD><CODE>preRNA</CODE> - Array containing all values of pre-existing RNA<DD><CODE>totalDescr</CODE> - A description that matches the columns of the arrays containing the values to the labels of the columns for total RNA<DD><CODE>newDescr</CODE> - matches columns of arrays to labels for newly transcribed RNA<DD><CODE>preDescr</CODE> - matches columns of arrays to labels for pre-existing RNA<DD><CODE>map</CODE> - A Map matching SpotId to index of RNA-arrays, containing only filtered values</DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="readInput(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List)"><!-- --></A><H3>
readInput</H3>
<PRE>
public void <B>readInput</B>(java.lang.String&nbsp;file,
                      java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
                      java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
                      java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
                      java.util.List&lt;java.lang.String&gt;&nbsp;addit)</PRE>
<DL>
<DD>Takes an input file, reads the values and saves them in the class variables of the Data object
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>file</CODE> - Name of the input file<DD><CODE>columnsTotal</CODE> - List containing labels of columns for total RNA<DD><CODE>columnsNew</CODE> - List containing labels of columns for newly transcribed RNA<DD><CODE>columnsPre</CODE> - List containing labels of columns for pre-existing RNA<DD><CODE>addit</CODE> - List containing labels of columns with additional information; null if no additional information is wanted</DL>
</DD>
</DL>
<HR>

<A NAME="addAttributes(java.lang.String)"><!-- --></A><H3>
addAttributes</H3>
<PRE>
public void <B>addAttributes</B>(java.lang.String&nbsp;inputfile)</PRE>
<DL>
<DD>Reads in a file containing two columns: column 1: spotID, column2: attribute (e.g. gene name)
 The resulting Mapping is added to the List attrMaps and accessible over the index received from attrMapsDescr,
 where the name of the column is stored
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>inputfile</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="writeOutput(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List)"><!-- --></A><H3>
writeOutput</H3>
<PRE>
public void <B>writeOutput</B>(java.lang.String&nbsp;output,
                        java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
                        java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
                        java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
                        java.util.List&lt;java.lang.String&gt;&nbsp;addit)</PRE>
<DL>
<DD>desired columns are written to a file; former order is not kept, instead all values of one type are added sequentially
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>output</CODE> - Name of the output file<DD><CODE>columnsTotal</CODE> - List containing labels of columns for total RNA that shall be exported<DD><CODE>columnsNew</CODE> - List containing labels of columns for newly transcribed RNA that shall be exported<DD><CODE>columnsPre</CODE> - List containing labels of columns for pre-existing RNA that shall be exported</DL>
</DD>
</DL>
<HR>

<A NAME="copyMap()"><!-- --></A><H3>
copyMap</H3>
<PRE>
public <A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt; <B>copyMap</B>()</PRE>
<DL>
<DD>Copies the content of the Mapping object
<P>
<DD><DL>

<DT><B>Returns:</B><DD>copy of the map that maps SpotID to indices</DL>
</DD>
</DL>
<HR>

<A NAME="calculateRatio(double[][], double[][])"><!-- --></A><H3>
calculateRatio</H3>
<PRE>
public double[] <B>calculateRatio</B>(double[][]&nbsp;rna,
                               double[][]&nbsp;total)</PRE>
<DL>
<DD>Calculates filtered Ratio of a set of RNA values (e.g. newly transcribed RNA) to a second set of values (e.g. total RNA)
 based on the specified method to calculate the average of multiple replicates. If none specified the default
 method is 'calculateRatioFirst'
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>rna</CODE> - A set of RNA values serving as numerator<DD><CODE>total</CODE> - A second set of RNA values serving as denominator
<DT><B>Returns:</B><DD>The ratio of the two averaged sets of RNA</DL>
</DD>
</DL>
<HR>

<A NAME="calculateNewMapping()"><!-- --></A><H3>
calculateNewMapping</H3>
<PRE>
public <A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt; <B>calculateNewMapping</B>()</PRE>
<DL>
<DD>Calculates the Mapping corresponding to the ratio
<P>
<DD><DL>

<DT><B>Returns:</B><DD>Mapping which matches SpotID to index of ratio array</DL>
</DD>
</DL>
<HR>

<A NAME="restrictRatio(double[][], int)"><!-- --></A><H3>
restrictRatio</H3>
<PRE>
public double[] <B>restrictRatio</B>(double[][]&nbsp;input,
                              int&nbsp;replicate)</PRE>
<DL>
<DD>Restricts a multidimensional array onto an array with one dimension containing the wanted replicate
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>input</CODE> - Ratio or Half life array
<DT><B>Returns:</B><DD>single dimensional array</DL>
</DD>
</DL>
<HR>

<A NAME="calculateRatioFirst(double[][])"><!-- --></A><H3>
calculateRatioFirst</H3>
<PRE>
public double[] <B>calculateRatioFirst</B>(double[][]&nbsp;allRatio)</PRE>
<DL>
<DD>calculates the average values from all replicates for the linear regression
 calculates the average of the ratios (RATIOFIRST method)
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>allRatio</CODE> - The array containing the previously calculated ratios
<DT><B>Returns:</B><DD>The average of the different ratios</DL>
</DD>
</DL>
<HR>

<A NAME="calculateAverageFirst(double[][], double[][])"><!-- --></A><H3>
calculateAverageFirst</H3>
<PRE>
public double[] <B>calculateAverageFirst</B>(double[][]&nbsp;rna,
                                      double[][]&nbsp;total)</PRE>
<DL>
<DD>calculates the average values from all replicates for the linear regression
 calculates the ratio of the averages (AVERAGEFIRST method)
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>rna</CODE> - A set of RNA that serves as numerator<DD><CODE>total</CODE> - A set of RNA that serves as denominator
<DT><B>Returns:</B><DD>The ratio of the averages of the given sets</DL>
</DD>
</DL>
<HR>

<A NAME="writeToFileArray(java.lang.String, double[])"><!-- --></A><H3>
writeToFileArray</H3>
<PRE>
public void <B>writeToFileArray</B>(java.lang.String&nbsp;output,
                             double[]&nbsp;array)</PRE>
<DL>
<DD>Writes the half life values contained in the object and in the map together with corresponding spotIds into a file
 Appropriate for half lives, ratios & pqs (for which the new mapping applies)
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>output</CODE> - Name of output file</DL>
</DD>
</DL>
<HR>

<A NAME="evaluate(java.lang.String, int)"><!-- --></A><H3>
evaluate</H3>
<PRE>
public double <B>evaluate</B>(java.lang.String&nbsp;fasta,
                       int&nbsp;column)</PRE>
<DL>
<DD>Method for the Uracil-based evaluation of the values
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>fasta</CODE> - The fasta file which contains the sequences<DD><CODE>column</CODE> - The number of the column (starting with 1) of the header in which
 the genename is contained (i.e. header: > genename | attribute | attribute leads to column = 1)
<DT><B>Returns:</B><DD>result Average uracil number - average log(e'/n')</DL>
</DD>
</DL>
<HR>

<A NAME="setMethod(int)"><!-- --></A><H3>
setMethod</H3>
<PRE>
public void <B>setMethod</B>(int&nbsp;i)</PRE>
<DL>
<DD>Sets the method used for calculating the average ratios over all replicates
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>i</CODE> - The integer value for the method</DL>
</DD>
</DL>
<HR>

<A NAME="setMap(data.Mapping)"><!-- --></A><H3>
setMap</H3>
<PRE>
public void <B>setMap</B>(<A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;&nbsp;map)</PRE>
<DL>
<DD>Sets the mapping corresponding to the whole data and filtered values
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>map</CODE> - The mapping corresponding to data</DL>
</DD>
</DL>
<HR>

<A NAME="setFilename(java.lang.String)"><!-- --></A><H3>
setFilename</H3>
<PRE>
public void <B>setFilename</B>(java.lang.String&nbsp;filename)</PRE>
<DL>
<DD>Sets the name of the file holding the data information
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>filename</CODE> - Name of the file with the data</DL>
</DD>
</DL>
<HR>

<A NAME="setAttributes(java.util.ArrayList)"><!-- --></A><H3>
setAttributes</H3>
<PRE>
public void <B>setAttributes</B>(java.util.ArrayList&lt;<A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.String&gt;&gt;&nbsp;attributes)</PRE>
<DL>
<DD>Sets the attributes list for the data
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>attributes</CODE> - The list of attribute mappings</DL>
</DD>
</DL>
<HR>

<A NAME="setAttrDescr(java.lang.String[])"><!-- --></A><H3>
setAttrDescr</H3>
<PRE>
public void <B>setAttrDescr</B>(java.lang.String[]&nbsp;attrDescr)</PRE>
<DL>
<DD>Sets the description of all attributes
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>attrDescr</CODE> - The description of attributes</DL>
</DD>
</DL>
<HR>

<A NAME="setAbsent(java.lang.String)"><!-- --></A><H3>
setAbsent</H3>
<PRE>
public static void <B>setAbsent</B>(java.lang.String&nbsp;absent)</PRE>
<DL>
<DD>Sets the name of the column (column label) which holds information on present/absent calls
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>absent</CODE> - Column label for present/absent calls</DL>
</DD>
</DL>
<HR>

<A NAME="setGeneName(java.lang.String)"><!-- --></A><H3>
setGeneName</H3>
<PRE>
public static void <B>setGeneName</B>(java.lang.String&nbsp;geneName)</PRE>
<DL>
<DD>Sets the name of the column which holds information on the gene name
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>geneName</CODE> - Column label for gene name</DL>
</DD>
</DL>
<HR>

<A NAME="setUnequal(boolean)"><!-- --></A><H3>
setUnequal</H3>
<PRE>
public void <B>setUnequal</B>(boolean&nbsp;unequal)</PRE>
<DL>
<DD>Set the value of the unequal variable, meaning if there is an
 unequal number of measurements (new, pre, total)
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>unequal</CODE> - True, if there is an unequal number of measurements</DL>
</DD>
</DL>
<HR>

<A NAME="setPqs(double[])"><!-- --></A><H3>
setPqs</H3>
<PRE>
public void <B>setPqs</B>(double[]&nbsp;pqs)</PRE>
<DL>
<DD>Sets the probe set quality scores
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>pqs</CODE> - The probe set quality scores</DL>
</DD>
</DL>
<HR>

<A NAME="getPqs()"><!-- --></A><H3>
getPqs</H3>
<PRE>
public double[] <B>getPqs</B>()</PRE>
<DL>
<DD>Returns the probe set quality scores
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The probe set quality scores</DL>
</DD>
</DL>
<HR>

<A NAME="getFilename()"><!-- --></A><H3>
getFilename</H3>
<PRE>
public java.lang.String <B>getFilename</B>()</PRE>
<DL>
<DD>Returns the name of the file containing the data
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The name of the data file</DL>
</DD>
</DL>
<HR>

<A NAME="getAttributes()"><!-- --></A><H3>
getAttributes</H3>
<PRE>
public java.util.ArrayList&lt;<A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.String&gt;&gt; <B>getAttributes</B>()</PRE>
<DL>
<DD>Returns the attribute lists
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The list of attribute mappings</DL>
</DD>
</DL>
<HR>

<A NAME="getMap()"><!-- --></A><H3>
getMap</H3>
<PRE>
public <A HREF="../data/Mapping.html" title="class in data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt; <B>getMap</B>()</PRE>
<DL>
<DD>Returns the map corresponding to the data
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The map for the data</DL>
</DD>
</DL>
<HR>

<A NAME="getAttrDescr()"><!-- --></A><H3>
getAttrDescr</H3>
<PRE>
public java.lang.String[] <B>getAttrDescr</B>()</PRE>
<DL>
<DD>Returns the description of the attributes
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The description of attributes</DL>
</DD>
</DL>
<HR>

<A NAME="getNewDescr()"><!-- --></A><H3>
getNewDescr</H3>
<PRE>
public java.lang.String[] <B>getNewDescr</B>()</PRE>
<DL>
<DD>Returns the description = column labels of newly transcribed RNA
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The description of pre RNA</DL>
</DD>
</DL>
<HR>

<A NAME="getNewRNA()"><!-- --></A><H3>
getNewRNA</H3>
<PRE>
public double[][] <B>getNewRNA</B>()</PRE>
<DL>
<DD>Returns all the values for newly transcribed RNA
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The measured values for newly transcribed RNA</DL>
</DD>
</DL>
<HR>

<A NAME="getPreDescr()"><!-- --></A><H3>
getPreDescr</H3>
<PRE>
public java.lang.String[] <B>getPreDescr</B>()</PRE>
<DL>
<DD>Returns the description = column labels of pre-existing RNA
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The description of pre-existing RNA</DL>
</DD>
</DL>
<HR>

<A NAME="getPreRNA()"><!-- --></A><H3>
getPreRNA</H3>
<PRE>
public double[][] <B>getPreRNA</B>()</PRE>
<DL>
<DD>Returns all the values for pre-existing RNA
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The measured values for pre-existing RNA</DL>
</DD>
</DL>
<HR>

<A NAME="getTotalDescr()"><!-- --></A><H3>
getTotalDescr</H3>
<PRE>
public java.lang.String[] <B>getTotalDescr</B>()</PRE>
<DL>
<DD>Returns the description = column labeles of total RNA
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The description of total RNA</DL>
</DD>
</DL>
<HR>

<A NAME="getTotalRNA()"><!-- --></A><H3>
getTotalRNA</H3>
<PRE>
public double[][] <B>getTotalRNA</B>()</PRE>
<DL>
<DD>Returns all the values for total RNA
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The measured values for total RNA</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberExperiments()"><!-- --></A><H3>
getNumberExperiments</H3>
<PRE>
public int <B>getNumberExperiments</B>()</PRE>
<DL>
<DD>Returns the number of measured experiments (not changeable through filtering!)
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The number of measured experiments</DL>
</DD>
</DL>
<HR>

<A NAME="getAvgLo()"><!-- --></A><H3>
getAvgLo</H3>
<PRE>
public double <B>getAvgLo</B>()</PRE>
<DL>
<DD>Returns the average log(e'/n')
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The average log(e'/n')</DL>
</DD>
</DL>
<HR>

<A NAME="getAvgUr()"><!-- --></A><H3>
getAvgUr</H3>
<PRE>
public double <B>getAvgUr</B>()</PRE>
<DL>
<DD>Returns the average number of Uracil
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The average number of Uracil</DL>
</DD>
</DL>
<HR>

<A NAME="getPlotName()"><!-- --></A><H3>
getPlotName</H3>
<PRE>
public java.lang.String <B>getPlotName</B>()</PRE>
<DL>
<DD>Returns the name of the file into which evaluation results will be printed
 as a plotable table
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The name of the file containing the plot information for evaluation</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/Data.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../data/Filter.html" title="class in data"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../index.html?data/Data.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Data.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
