<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_16) on Fri Dec 18 11:27:07 CET 2009 -->
<TITLE>
Data
</TITLE>

<META NAME="date" CONTENT="2009-12-18">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="Data";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/Data.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../halo/data/Filter.html" title="class in halo.data"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?halo/data/Data.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Data.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
halo.data</FONT>
<BR>
Class Data</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../resources/inherit.gif" ALT="extended by "><B>halo.data.Data</B>
</PRE>
<HR>
<DL>
<DT><PRE>public class <B>Data</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
Reading and containing RNA values from an input file and providing methods
 for the calculation of RNA ratios as well as methods to save the filtered
 values to a file.
<P>

<P>
<DL>
<DT><B>Author:</B></DT>
  <DD>Stefanie Kaufmann</DD>
</DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#AVERAGEFIRST">AVERAGEFIRST</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#GENEDEFAULT">GENEDEFAULT</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#LOGEN">LOGEN</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#LOGEU">LOGEU</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#LOGUN">LOGUN</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#RATIOFIRST">RATIOFIRST</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#REPLICATE">REPLICATE</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../halo/data/Data.html#Data()">Data</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs an empty Data-Object, so that parameters for data
 loading can be set before starting.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../halo/data/Data.html#Data(double[][], double[][], double[][], java.lang.String[], java.lang.String[], java.lang.String[], halo.data.Mapping)">Data</A></B>(double[][]&nbsp;totalRNA,
     double[][]&nbsp;newRNA,
     double[][]&nbsp;preRNA,
     java.lang.String[]&nbsp;totalDescr,
     java.lang.String[]&nbsp;newDescr,
     java.lang.String[]&nbsp;preDescr,
     <A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;&nbsp;map)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a Data-Object based on already existing data.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../halo/data/Data.html#Data(java.lang.String, java.util.List, java.util.List, java.util.List, boolean)">Data</A></B>(java.lang.String&nbsp;file,
     java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
     java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
     java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
     boolean&nbsp;log)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a Data-Object based on an Input-File and the list of
 columns to be read with no additional attributes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../halo/data/Data.html#Data(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List, boolean)">Data</A></B>(java.lang.String&nbsp;file,
     java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
     java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
     java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
     java.util.List&lt;java.lang.String&gt;&nbsp;attr,
     boolean&nbsp;log)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a Data-Object based on an Input-File and the list of
 columns to be read, defines columns containing attributes.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#addAttributes(java.lang.String)">addAttributes</A></B>(java.lang.String&nbsp;inputfile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads in every attribute from a given file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#addAttributes(java.lang.String, java.util.ArrayList)">addAttributes</A></B>(java.lang.String&nbsp;inputfile,
              java.util.ArrayList&lt;java.lang.String&gt;&nbsp;columnLabels)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads in regular attributes from a file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#addAttributes(java.lang.String, java.util.ArrayList, boolean)">addAttributes</A></B>(java.lang.String&nbsp;inputfile,
              java.util.ArrayList&lt;java.lang.String&gt;&nbsp;columnLabels,
              boolean&nbsp;turnReverseMapOff)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads in a file containing several columns, of which one contains
 the spotID, the others contain attributes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#addPresenceCalls(java.lang.String, java.util.ArrayList)">addPresenceCalls</A></B>(java.lang.String&nbsp;inputfile,
                 java.util.ArrayList&lt;java.lang.String&gt;&nbsp;columnLabels)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads in attributes from a file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#calculateAverageFirst(double[][], double[][], java.util.HashMap)">calculateAverageFirst</A></B>(double[][]&nbsp;rna,
                      double[][]&nbsp;total,
                      java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corr)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculates the average values from all replicates for the
 linear regression through calculating the ratio of the
 averages (AVERAGEFIRST method).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#calculateMappingForAttributes(java.lang.String[][], java.lang.String[])">calculateMappingForAttributes</A></B>(java.lang.String[][]&nbsp;allAttributes,
                              java.lang.String[]&nbsp;ids)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Transcribes the attributes into a mapping id -> attribute.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#calculateNewMapping()">calculateNewMapping</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculates the Mapping corresponding to the ratio.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#calculateRatio(double[][], double[][], java.util.HashMap)">calculateRatio</A></B>(double[][]&nbsp;rna,
               double[][]&nbsp;total,
               java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corr)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculates the ratios for an average over a set of replicates.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#calculateRatio(double[][], double[][], java.util.HashMap, int)">calculateRatio</A></B>(double[][]&nbsp;rna,
               double[][]&nbsp;total,
               java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corr,
               int&nbsp;replicate)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculates filtered Ratio of a set of RNA values (e.g. newly
 transcribed RNA) to a second set of values (e.g. total RNA)
 based on the specified method to calculate the average of
 multiple replicates.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#calculateRatioFirst(double[][])">calculateRatioFirst</A></B>(double[][]&nbsp;allRatio)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;calculates the average values from all replicates for the
 linear regression.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#copyMap()">copyMap</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Copies the content of the Mapping object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#evaluate(java.lang.String, int, java.lang.String, java.lang.String, java.util.HashMap, boolean)">evaluate</A></B>(java.lang.String&nbsp;fasta,
         int&nbsp;column,
         java.lang.String&nbsp;uraMeth,
         java.lang.String&nbsp;printName,
         java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corr,
         boolean&nbsp;plot)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Method for the Uracil-based evaluation of the values: Compares
 (average) uracil number of the sequence to logarithm of the ratio.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#findMiniMaxi()">findMiniMaxi</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the new minimum and maximum</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getAbsent()">getAbsent</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the call used for present/absent filtering.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getAbsentLabel()">getAbsentLabel</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the name of the column with present/absent calls.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getAttrDescr()">getAttrDescr</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the description of the attributes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.ArrayList&lt;<A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.String&gt;&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getAttributes()">getAttributes</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the attribute lists.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getAvgLo()">getAvgLo</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the average log(e'/n').</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getAvgUr()">getAvgUr</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the average number of Uracil.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getCorrNewPre()">getCorrNewPre</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the correction bias values for the ratio newly transcribed/pre-existing.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getCorrNewTot()">getCorrNewTot</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the correction bias values for the ratio newly transcribed/total.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getCorrPreTot()">getCorrPreTot</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the correction bias values for the ratio pre-existing/total.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getDat()">getDat</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the evaluated normalized data.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getFilename()">getFilename</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the name of the file containing the data.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getGeneName()">getGeneName</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the name of the column with gene names.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getMap()">getMap</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the map corresponding to the data.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getMaximum()">getMaximum</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the maximal expression value present in the data</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getMethod()">getMethod</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the number of the method used for ratio calculation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getMinimum()">getMinimum</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the minimal expression value present in the data</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getNewDescr()">getNewDescr</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the description = column labels of newly transcribed RNA.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getNewRNA()">getNewRNA</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns all the values for newly transcribed RNA.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getNumberExperiments()">getNumberExperiments</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the number of measured experiments
 (not changeable through filtering!).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getPlotName()">getPlotName</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the name of the file into which evaluation
 results will be printed as a plottable table.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getPqs()">getPqs</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the probe set quality scores.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getPreDescr()">getPreDescr</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the description = column labels of pre-existing RNA.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getPreRNA()">getPreRNA</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns all the values for pre-existing RNA.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getReplicate()">getReplicate</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the number of the replicate used for calculations.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getReplicatesTotal()">getReplicatesTotal</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the total number of replicates.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getSize()">getSize</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the number of probesets present after filtering.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getSpot()">getSpot</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the list of spots.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getSpotid()">getSpotid</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the name of the column with probeset ids.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getTotalDescr()">getTotalDescr</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the description = column labeles of total RNA.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#getTotalRNA()">getTotalRNA</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns all the values for total RNA.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#isUnequal()">isUnequal</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns TRUE if unequal number of RNA measurements has been loaded.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#loadAttributesFromDatafile(java.util.ArrayList)">loadAttributesFromDatafile</A></B>(java.util.ArrayList&lt;java.lang.String&gt;&nbsp;newAttributes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Loads attributes from original data file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#loadAttributesFromDatafile(java.util.ArrayList, boolean)">loadAttributesFromDatafile</A></B>(java.util.ArrayList&lt;java.lang.String&gt;&nbsp;newAttributes,
                           boolean&nbsp;turnReverseMapOff)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Loads attributes from the original data file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#loadData(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List, boolean)">loadData</A></B>(java.lang.String&nbsp;file,
         java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
         java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
         java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
         java.util.List&lt;java.lang.String&gt;&nbsp;attr,
         boolean&nbsp;log)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Loads the data from a file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#loadPresentCallsFromDatafile(java.util.ArrayList)">loadPresentCallsFromDatafile</A></B>(java.util.ArrayList&lt;java.lang.String&gt;&nbsp;presentCallAttributes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Loads present calls or other attributes with a small number of
 values from the original data file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#readInput(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List, boolean)">readInput</A></B>(java.lang.String&nbsp;file,
          java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
          java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
          java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
          java.util.List&lt;java.lang.String&gt;&nbsp;addit,
          boolean&nbsp;log)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Takes an input file, reads the values and saves them in the class
 variables of the Data object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#restrictRatio(double[][], int)">restrictRatio</A></B>(double[][]&nbsp;input,
              int&nbsp;replicate)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Restricts a multidimensional array onto an array with one
 dimension containing the wanted replicate, while keeping
 the order.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setAbsent(java.lang.String)">setAbsent</A></B>(java.lang.String&nbsp;absent)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the name of the column (column label) which holds information
 on present/absent.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setAttrDescr(java.lang.String[])">setAttrDescr</A></B>(java.lang.String[]&nbsp;attrDescr)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the description of all attributes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setAttributes(java.util.ArrayList)">setAttributes</A></B>(java.util.ArrayList&lt;<A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.String&gt;&gt;&nbsp;attributes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the attributes list for the data.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setCorrNewPre(java.util.HashMap)">setCorrNewPre</A></B>(java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corr)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the correction bias values for the ratio newly transcribed/pre-existing.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setCorrNewTot(java.util.HashMap)">setCorrNewTot</A></B>(java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corrNewTot)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the correction bias values for the ratio newly transcribed/total.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setCorrPreTot(java.util.HashMap)">setCorrPreTot</A></B>(java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corrPreTot)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the correction bias values for the ratio pre-existing/total.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setFilename(java.lang.String)">setFilename</A></B>(java.lang.String&nbsp;filename)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the name of the file holding the data information.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setGeneName(java.lang.String)">setGeneName</A></B>(java.lang.String&nbsp;geneName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the name of the column which holds information on the gene name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setMap(halo.data.Mapping)">setMap</A></B>(<A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;&nbsp;map)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the mapping corresponding to the whole data and filtered values.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setMaximum(double)">setMaximum</A></B>(double&nbsp;maximum)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the maximum RNA value</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setMethod(int)">setMethod</A></B>(int&nbsp;i)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the method used for calculating the average ratios
 over all replicates.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setMinimum(double)">setMinimum</A></B>(double&nbsp;minimum)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the minimum RNA value</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setPqs(double[])">setPqs</A></B>(double[]&nbsp;pqs)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the probe set quality scores.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setReplicate(int)">setReplicate</A></B>(int&nbsp;replicate)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the replicate used for ratio calculations, if method is
 restricting of ratios to one replicate.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setSpotid(java.lang.String)">setSpotid</A></B>(java.lang.String&nbsp;probesetId)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the name of the column with probeset ids.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#setUnequal(boolean)">setUnequal</A></B>(boolean&nbsp;unequal)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Set the value of the unequal variable, meaning if there is an
 unequal number of measurements (new, pre, total).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#writeOutput(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List)">writeOutput</A></B>(java.lang.String&nbsp;output,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
            java.util.List&lt;java.lang.String&gt;&nbsp;addit)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Desired columns are written to a file; former order is not kept,
 instead all values of one type are added sequentially.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../halo/data/Data.html#writeToFileArray(java.lang.String, double[])">writeToFileArray</A></B>(java.lang.String&nbsp;output,
                 double[]&nbsp;array)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes the half-life values contained in the array and in the
 map together with corresponding spotIds into a file.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="RATIOFIRST"><!-- --></A><H3>
RATIOFIRST</H3>
<PRE>
public static final int <B>RATIOFIRST</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#halo.data.Data.RATIOFIRST">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="AVERAGEFIRST"><!-- --></A><H3>
AVERAGEFIRST</H3>
<PRE>
public static final int <B>AVERAGEFIRST</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#halo.data.Data.AVERAGEFIRST">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="REPLICATE"><!-- --></A><H3>
REPLICATE</H3>
<PRE>
public static final int <B>REPLICATE</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#halo.data.Data.REPLICATE">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="LOGEN"><!-- --></A><H3>
LOGEN</H3>
<PRE>
public static final java.lang.String <B>LOGEN</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#halo.data.Data.LOGEN">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="LOGUN"><!-- --></A><H3>
LOGUN</H3>
<PRE>
public static final java.lang.String <B>LOGUN</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#halo.data.Data.LOGUN">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="LOGEU"><!-- --></A><H3>
LOGEU</H3>
<PRE>
public static final java.lang.String <B>LOGEU</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#halo.data.Data.LOGEU">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="GENEDEFAULT"><!-- --></A><H3>
GENEDEFAULT</H3>
<PRE>
public static final java.lang.String <B>GENEDEFAULT</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#halo.data.Data.GENEDEFAULT">Constant Field Values</A></DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="Data(java.lang.String, java.util.List, java.util.List, java.util.List, boolean)"><!-- --></A><H3>
Data</H3>
<PRE>
public <B>Data</B>(java.lang.String&nbsp;file,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
            boolean&nbsp;log)</PRE>
<DL>
<DD>Constructs a Data-Object based on an Input-File and the list of
 columns to be read with no additional attributes.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>file</CODE> - Name of the Input-file<DD><CODE>columnsTotal</CODE> - List containing the labels of columns for
 total RNA<DD><CODE>columnsNew</CODE> - List containing the labels of columns for newly
 transcribed RNA<DD><CODE>columnsPre</CODE> - List containing the labels of columns for
 pre-existing RNA</DL>
</DL>
<HR>

<A NAME="Data(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List, boolean)"><!-- --></A><H3>
Data</H3>
<PRE>
public <B>Data</B>(java.lang.String&nbsp;file,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
            java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
            java.util.List&lt;java.lang.String&gt;&nbsp;attr,
            boolean&nbsp;log)</PRE>
<DL>
<DD>Constructs a Data-Object based on an Input-File and the list of
 columns to be read, defines columns containing attributes.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>file</CODE> - Name of the Input-file<DD><CODE>columnsTotal</CODE> - List containing the labels of columns for
 total RNA<DD><CODE>columnsNew</CODE> - List containing the labels of columns for
 newly transcribed RNA<DD><CODE>columnsPre</CODE> - List containing the labels of columns for
 pre-existing RNA<DD><CODE>attr</CODE> - List containing the labels of columns for attributes</DL>
</DL>
<HR>

<A NAME="Data()"><!-- --></A><H3>
Data</H3>
<PRE>
public <B>Data</B>()</PRE>
<DL>
<DD>Constructs an empty Data-Object, so that parameters for data
 loading can be set before starting.
<P>
</DL>
<HR>

<A NAME="Data(double[][], double[][], double[][], java.lang.String[], java.lang.String[], java.lang.String[], halo.data.Mapping)"><!-- --></A><H3>
Data</H3>
<PRE>
public <B>Data</B>(double[][]&nbsp;totalRNA,
            double[][]&nbsp;newRNA,
            double[][]&nbsp;preRNA,
            java.lang.String[]&nbsp;totalDescr,
            java.lang.String[]&nbsp;newDescr,
            java.lang.String[]&nbsp;preDescr,
            <A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;&nbsp;map)</PRE>
<DL>
<DD>Constructs a Data-Object based on already existing data.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>totalRNA</CODE> - Array containing all values of total RNA<DD><CODE>newRNA</CODE> - Array containing all values of newly transcribed RNA<DD><CODE>preRNA</CODE> - Array containing all values of pre-existing RNA<DD><CODE>totalDescr</CODE> - A description that matches the columns of the
 arrays containing the values to the labels of the columns for
 total RNA<DD><CODE>newDescr</CODE> - matches columns of arrays to labels for newly
 transcribed RNA<DD><CODE>preDescr</CODE> - matches columns of arrays to labels for pre-existing
 RNA<DD><CODE>map</CODE> - A Map matching SpotId to index of RNA-arrays, containing
 only filtered values</DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="loadData(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List, boolean)"><!-- --></A><H3>
loadData</H3>
<PRE>
public void <B>loadData</B>(java.lang.String&nbsp;file,
                     java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
                     java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
                     java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
                     java.util.List&lt;java.lang.String&gt;&nbsp;attr,
                     boolean&nbsp;log)</PRE>
<DL>
<DD>Loads the data from a file.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>file</CODE> - Name of the Input-file<DD><CODE>columnsTotal</CODE> - List containing the labels of columns for
 total RNA<DD><CODE>columnsNew</CODE> - List containing the labels of columns for
 newly transcribed RNA<DD><CODE>columnsPre</CODE> - List containing the labels of columns for
 pre-existing RNA<DD><CODE>attr</CODE> - List containing the labels of columns for attributes</DL>
</DD>
</DL>
<HR>

<A NAME="readInput(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List, boolean)"><!-- --></A><H3>
readInput</H3>
<PRE>
public void <B>readInput</B>(java.lang.String&nbsp;file,
                      java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
                      java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
                      java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
                      java.util.List&lt;java.lang.String&gt;&nbsp;addit,
                      boolean&nbsp;log)</PRE>
<DL>
<DD>Takes an input file, reads the values and saves them in the class
 variables of the Data object.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>file</CODE> - Name of the input file<DD><CODE>columnsTotal</CODE> - List containing labels of columns for total RNA<DD><CODE>columnsNew</CODE> - List containing labels of columns for newly
 transcribed RNA<DD><CODE>columnsPre</CODE> - List containing labels of columns for pre-existing
 RNA<DD><CODE>addit</CODE> - List containing labels of columns with additional
 information; null if no additional information is wanted</DL>
</DD>
</DL>
<HR>

<A NAME="calculateMappingForAttributes(java.lang.String[][], java.lang.String[])"><!-- --></A><H3>
calculateMappingForAttributes</H3>
<PRE>
public void <B>calculateMappingForAttributes</B>(java.lang.String[][]&nbsp;allAttributes,
                                          java.lang.String[]&nbsp;ids)</PRE>
<DL>
<DD>Transcribes the attributes into a mapping id -> attribute.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>allAttributes</CODE> - A list of attributes for each id<DD><CODE>ids</CODE> - The ids corresponding to the indices</DL>
</DD>
</DL>
<HR>

<A NAME="addAttributes(java.lang.String)"><!-- --></A><H3>
addAttributes</H3>
<PRE>
public void <B>addAttributes</B>(java.lang.String&nbsp;inputfile)</PRE>
<DL>
<DD>Reads in every attribute from a given file.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>inputfile</CODE> - Name of the file that contains attributes</DL>
</DD>
</DL>
<HR>

<A NAME="addAttributes(java.lang.String, java.util.ArrayList)"><!-- --></A><H3>
addAttributes</H3>
<PRE>
public void <B>addAttributes</B>(java.lang.String&nbsp;inputfile,
                          java.util.ArrayList&lt;java.lang.String&gt;&nbsp;columnLabels)</PRE>
<DL>
<DD>Reads in regular attributes from a file.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>inputfile</CODE> - Name of the file that contains attributes<DD><CODE>columnLabels</CODE> - Labels of relevant columns</DL>
</DD>
</DL>
<HR>

<A NAME="addPresenceCalls(java.lang.String, java.util.ArrayList)"><!-- --></A><H3>
addPresenceCalls</H3>
<PRE>
public void <B>addPresenceCalls</B>(java.lang.String&nbsp;inputfile,
                             java.util.ArrayList&lt;java.lang.String&gt;&nbsp;columnLabels)</PRE>
<DL>
<DD>Reads in attributes from a file. Specific for attributes with a
 very small number of values (e.g. present calls) to speed up
 the loading. Please note that with this method the reverse mapping
 from attribute to list of IDs is not possible.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>inputfile</CODE> - Name of the file that contains attributes<DD><CODE>columnLabels</CODE> - Labels of relevant columns</DL>
</DD>
</DL>
<HR>

<A NAME="addAttributes(java.lang.String, java.util.ArrayList, boolean)"><!-- --></A><H3>
addAttributes</H3>
<PRE>
public void <B>addAttributes</B>(java.lang.String&nbsp;inputfile,
                          java.util.ArrayList&lt;java.lang.String&gt;&nbsp;columnLabels,
                          boolean&nbsp;turnReverseMapOff)</PRE>
<DL>
<DD>Reads in a file containing several columns, of which one contains
 the spotID, the others contain attributes.
 The resulting Mappings are added to the List attrMaps and accessible
 over the index received from attrMapsDescr, where the name of the
 column is stored
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>inputfile</CODE> - Name of the file that contains attributes<DD><CODE>columnLabels</CODE> - Labels of relevant columns<DD><CODE>turnReverseMapOff</CODE> - TRUE it attributes are present calls or
 similar attributes with a very small number of values</DL>
</DD>
</DL>
<HR>

<A NAME="loadAttributesFromDatafile(java.util.ArrayList)"><!-- --></A><H3>
loadAttributesFromDatafile</H3>
<PRE>
public void <B>loadAttributesFromDatafile</B>(java.util.ArrayList&lt;java.lang.String&gt;&nbsp;newAttributes)</PRE>
<DL>
<DD>Loads attributes from original data file.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>newAttributes</CODE> - The list of attributes that will be loaded</DL>
</DD>
</DL>
<HR>

<A NAME="loadPresentCallsFromDatafile(java.util.ArrayList)"><!-- --></A><H3>
loadPresentCallsFromDatafile</H3>
<PRE>
public void <B>loadPresentCallsFromDatafile</B>(java.util.ArrayList&lt;java.lang.String&gt;&nbsp;presentCallAttributes)</PRE>
<DL>
<DD>Loads present calls or other attributes with a small number of
 values from the original data file.
 Please note that reverse mapping (attribute to list of IDs) is
 not available for attributes loaded with this method for the
 sake of speed.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>presentCallAttributes</CODE> - The list of attributes that will
 be loaded</DL>
</DD>
</DL>
<HR>

<A NAME="loadAttributesFromDatafile(java.util.ArrayList, boolean)"><!-- --></A><H3>
loadAttributesFromDatafile</H3>
<PRE>
public void <B>loadAttributesFromDatafile</B>(java.util.ArrayList&lt;java.lang.String&gt;&nbsp;newAttributes,
                                       boolean&nbsp;turnReverseMapOff)</PRE>
<DL>
<DD>Loads attributes from the original data file.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>newAttributes</CODE> - The list of attributes that will be loaded<DD><CODE>turnReverseMapOff</CODE> - TRUE if attributes are present calls or
 similar attributes with very few possible values</DL>
</DD>
</DL>
<HR>

<A NAME="writeOutput(java.lang.String, java.util.List, java.util.List, java.util.List, java.util.List)"><!-- --></A><H3>
writeOutput</H3>
<PRE>
public void <B>writeOutput</B>(java.lang.String&nbsp;output,
                        java.util.List&lt;java.lang.String&gt;&nbsp;columnsTotal,
                        java.util.List&lt;java.lang.String&gt;&nbsp;columnsNew,
                        java.util.List&lt;java.lang.String&gt;&nbsp;columnsPre,
                        java.util.List&lt;java.lang.String&gt;&nbsp;addit)</PRE>
<DL>
<DD>Desired columns are written to a file; former order is not kept,
 instead all values of one type are added sequentially.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>output</CODE> - Name of the output file<DD><CODE>columnsTotal</CODE> - List containing labels of columns for total
 RNA that shall be exported<DD><CODE>columnsNew</CODE> - List containing labels of columns for newly
 transcribed RNA that shall be exported<DD><CODE>columnsPre</CODE> - List containing labels of columns for pre-existing
 RNA that shall be exported<DD><CODE>addit</CODE> - List containing labels of columns for attributes</DL>
</DD>
</DL>
<HR>

<A NAME="copyMap()"><!-- --></A><H3>
copyMap</H3>
<PRE>
public <A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt; <B>copyMap</B>()</PRE>
<DL>
<DD>Copies the content of the Mapping object.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>copy of the map that maps SpotID to indices</DL>
</DD>
</DL>
<HR>

<A NAME="calculateRatio(double[][], double[][], java.util.HashMap, int)"><!-- --></A><H3>
calculateRatio</H3>
<PRE>
public double[] <B>calculateRatio</B>(double[][]&nbsp;rna,
                               double[][]&nbsp;total,
                               java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corr,
                               int&nbsp;replicate)</PRE>
<DL>
<DD>Calculates filtered Ratio of a set of RNA values (e.g. newly
 transcribed RNA) to a second set of values (e.g. total RNA)
 based on the specified method to calculate the average of
 multiple replicates. If none specified the default method is
 'calculateRatioFirst'. Please perform the filtering method
 'filterCorrectionBias' before using this method.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>rna</CODE> - A set of RNA values serving as numerator<DD><CODE>total</CODE> - A second set of RNA values serving as denominator<DD><CODE>corr</CODE> - The map for ID and bias correction, null if no
 correction should take place<DD><CODE>replicate</CODE> - The number of the replicate (starting from 0)
 that will be used for method REPLICATE; -1 if average should
 be calculated
<DT><B>Returns:</B><DD>The ratio of the two averaged sets of RNA</DL>
</DD>
</DL>
<HR>

<A NAME="calculateRatio(double[][], double[][], java.util.HashMap)"><!-- --></A><H3>
calculateRatio</H3>
<PRE>
public double[] <B>calculateRatio</B>(double[][]&nbsp;rna,
                               double[][]&nbsp;total,
                               java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corr)</PRE>
<DL>
<DD>Calculates the ratios for an average over a set of replicates.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>rna</CODE> - A set of RNA values serving as numerator<DD><CODE>total</CODE> - A second set of RNA values serving as denominator<DD><CODE>corr</CODE> - The map for ID and bias correction, null if no
 correction should take place
<DT><B>Returns:</B><DD>The ratio of the two averaged sets of RNA</DL>
</DD>
</DL>
<HR>

<A NAME="calculateNewMapping()"><!-- --></A><H3>
calculateNewMapping</H3>
<PRE>
public <A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt; <B>calculateNewMapping</B>()</PRE>
<DL>
<DD>Calculates the Mapping corresponding to the ratio.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>Mapping which matches SpotID to index of ratio array</DL>
</DD>
</DL>
<HR>

<A NAME="restrictRatio(double[][], int)"><!-- --></A><H3>
restrictRatio</H3>
<PRE>
public double[] <B>restrictRatio</B>(double[][]&nbsp;input,
                              int&nbsp;replicate)</PRE>
<DL>
<DD>Restricts a multidimensional array onto an array with one
 dimension containing the wanted replicate, while keeping
 the order.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>input</CODE> - Ratio or half-life array<DD><CODE>replicate</CODE> - The replicate to which the ratio will be restricted
<DT><B>Returns:</B><DD>single dimensional array</DL>
</DD>
</DL>
<HR>

<A NAME="calculateRatioFirst(double[][])"><!-- --></A><H3>
calculateRatioFirst</H3>
<PRE>
public double[] <B>calculateRatioFirst</B>(double[][]&nbsp;allRatio)</PRE>
<DL>
<DD>calculates the average values from all replicates for the
 linear regression.
 calculates the average of the ratios (RATIOFIRST method)
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>allRatio</CODE> - The array containing the previously calculated ratios
<DT><B>Returns:</B><DD>The average of the different ratios</DL>
</DD>
</DL>
<HR>

<A NAME="calculateAverageFirst(double[][], double[][], java.util.HashMap)"><!-- --></A><H3>
calculateAverageFirst</H3>
<PRE>
public double[] <B>calculateAverageFirst</B>(double[][]&nbsp;rna,
                                      double[][]&nbsp;total,
                                      java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corr)</PRE>
<DL>
<DD>Calculates the average values from all replicates for the
 linear regression through calculating the ratio of the
 averages (AVERAGEFIRST method).
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>rna</CODE> - A set of RNA that serves as numerator<DD><CODE>total</CODE> - A set of RNA that serves as denominator<DD><CODE>corr</CODE> - The map for ID and bias correction, null if no
 correction should take place
<DT><B>Returns:</B><DD>The ratio of the averages of the given sets</DL>
</DD>
</DL>
<HR>

<A NAME="writeToFileArray(java.lang.String, double[])"><!-- --></A><H3>
writeToFileArray</H3>
<PRE>
public void <B>writeToFileArray</B>(java.lang.String&nbsp;output,
                             double[]&nbsp;array)</PRE>
<DL>
<DD>Writes the half-life values contained in the array and in the
 map together with corresponding spotIds into a file.
 Appropriate for half-lives, ratios & pqs (for which the new
 mapping applies)
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>output</CODE> - Name of output file<DD><CODE>array</CODE> - List containing any set of values</DL>
</DD>
</DL>
<HR>

<A NAME="evaluate(java.lang.String, int, java.lang.String, java.lang.String, java.util.HashMap, boolean)"><!-- --></A><H3>
evaluate</H3>
<PRE>
public double <B>evaluate</B>(java.lang.String&nbsp;fasta,
                       int&nbsp;column,
                       java.lang.String&nbsp;uraMeth,
                       java.lang.String&nbsp;printName,
                       java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corr,
                       boolean&nbsp;plot)</PRE>
<DL>
<DD>Method for the Uracil-based evaluation of the values: Compares
 (average) uracil number of the sequence to logarithm of the ratio.
 For multiple transcript entries per gene an average uracil number
 is calculated.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>fasta</CODE> - The fasta file which contains the sequences<DD><CODE>column</CODE> - The number of the column (starting with 1) of the
 header in which the genename is contained (i.e. header:
 > genename | attribute | attribute leads to column = 1)<DD><CODE>uraMeth</CODE> - Either "log(e'/n')", or "log(e'/u')", or
 "log(u'/n')", according to the ratio which will be evaluated<DD><CODE>printName</CODE> - Name of the output file for plotting information<DD><CODE>corr</CODE> - The map for ID and bias correction, null if no
 correction should take place<DD><CODE>plot</CODE> - TRUE if histogram plot is wanted, FALSE otherwise
<DT><B>Returns:</B><DD>result Average uracil number</DL>
</DD>
</DL>
<HR>

<A NAME="findMiniMaxi()"><!-- --></A><H3>
findMiniMaxi</H3>
<PRE>
public void <B>findMiniMaxi</B>()</PRE>
<DL>
<DD>Sets the new minimum and maximum
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setCorrNewPre(java.util.HashMap)"><!-- --></A><H3>
setCorrNewPre</H3>
<PRE>
public void <B>setCorrNewPre</B>(java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corr)</PRE>
<DL>
<DD>Sets the correction bias values for the ratio newly transcribed/pre-existing.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>corr</CODE> - The correction bias values for the ratio newly transcribed/pre-existing</DL>
</DD>
</DL>
<HR>

<A NAME="setCorrNewTot(java.util.HashMap)"><!-- --></A><H3>
setCorrNewTot</H3>
<PRE>
public void <B>setCorrNewTot</B>(java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corrNewTot)</PRE>
<DL>
<DD>Sets the correction bias values for the ratio newly transcribed/total.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>corrNewTot</CODE> - The correction bias values for the ratio
 newly transcribed/total</DL>
</DD>
</DL>
<HR>

<A NAME="setCorrPreTot(java.util.HashMap)"><!-- --></A><H3>
setCorrPreTot</H3>
<PRE>
public void <B>setCorrPreTot</B>(java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;corrPreTot)</PRE>
<DL>
<DD>Sets the correction bias values for the ratio pre-existing/total.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>corrPreTot</CODE> - The correction bias values for the ratio
 pre-existing/total</DL>
</DD>
</DL>
<HR>

<A NAME="setMethod(int)"><!-- --></A><H3>
setMethod</H3>
<PRE>
public void <B>setMethod</B>(int&nbsp;i)</PRE>
<DL>
<DD>Sets the method used for calculating the average ratios
 over all replicates.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>i</CODE> - The integer value for the method</DL>
</DD>
</DL>
<HR>

<A NAME="setMap(halo.data.Mapping)"><!-- --></A><H3>
setMap</H3>
<PRE>
public void <B>setMap</B>(<A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt;&nbsp;map)</PRE>
<DL>
<DD>Sets the mapping corresponding to the whole data and filtered values.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>map</CODE> - The mapping corresponding to data</DL>
</DD>
</DL>
<HR>

<A NAME="setMaximum(double)"><!-- --></A><H3>
setMaximum</H3>
<PRE>
public void <B>setMaximum</B>(double&nbsp;maximum)</PRE>
<DL>
<DD>Sets the maximum RNA value
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>maximum</CODE> - Maximum RNA value</DL>
</DD>
</DL>
<HR>

<A NAME="setMinimum(double)"><!-- --></A><H3>
setMinimum</H3>
<PRE>
public void <B>setMinimum</B>(double&nbsp;minimum)</PRE>
<DL>
<DD>Sets the minimum RNA value
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>minimum</CODE> - Minimum RNA value</DL>
</DD>
</DL>
<HR>

<A NAME="setFilename(java.lang.String)"><!-- --></A><H3>
setFilename</H3>
<PRE>
public void <B>setFilename</B>(java.lang.String&nbsp;filename)</PRE>
<DL>
<DD>Sets the name of the file holding the data information.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>filename</CODE> - Name of the file with the data</DL>
</DD>
</DL>
<HR>

<A NAME="setAttributes(java.util.ArrayList)"><!-- --></A><H3>
setAttributes</H3>
<PRE>
public void <B>setAttributes</B>(java.util.ArrayList&lt;<A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.String&gt;&gt;&nbsp;attributes)</PRE>
<DL>
<DD>Sets the attributes list for the data.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>attributes</CODE> - The list of attribute mappings</DL>
</DD>
</DL>
<HR>

<A NAME="setAttrDescr(java.lang.String[])"><!-- --></A><H3>
setAttrDescr</H3>
<PRE>
public void <B>setAttrDescr</B>(java.lang.String[]&nbsp;attrDescr)</PRE>
<DL>
<DD>Sets the description of all attributes.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>attrDescr</CODE> - The description of attributes</DL>
</DD>
</DL>
<HR>

<A NAME="setAbsent(java.lang.String)"><!-- --></A><H3>
setAbsent</H3>
<PRE>
public void <B>setAbsent</B>(java.lang.String&nbsp;absent)</PRE>
<DL>
<DD>Sets the name of the column (column label) which holds information
 on present/absent.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>absent</CODE> - Column label for present/absent calls</DL>
</DD>
</DL>
<HR>

<A NAME="setGeneName(java.lang.String)"><!-- --></A><H3>
setGeneName</H3>
<PRE>
public void <B>setGeneName</B>(java.lang.String&nbsp;geneName)</PRE>
<DL>
<DD>Sets the name of the column which holds information on the gene name.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>geneName</CODE> - Column label for gene name</DL>
</DD>
</DL>
<HR>

<A NAME="setSpotid(java.lang.String)"><!-- --></A><H3>
setSpotid</H3>
<PRE>
public void <B>setSpotid</B>(java.lang.String&nbsp;probesetId)</PRE>
<DL>
<DD>Sets the name of the column with probeset ids.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>probesetId</CODE> - Name of the column with probeset ids</DL>
</DD>
</DL>
<HR>

<A NAME="setUnequal(boolean)"><!-- --></A><H3>
setUnequal</H3>
<PRE>
public void <B>setUnequal</B>(boolean&nbsp;unequal)</PRE>
<DL>
<DD>Set the value of the unequal variable, meaning if there is an
 unequal number of measurements (new, pre, total).
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>unequal</CODE> - True, if there is an unequal number of measurements</DL>
</DD>
</DL>
<HR>

<A NAME="setReplicate(int)"><!-- --></A><H3>
setReplicate</H3>
<PRE>
public void <B>setReplicate</B>(int&nbsp;replicate)</PRE>
<DL>
<DD>Sets the replicate used for ratio calculations, if method is
 restricting of ratios to one replicate.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>replicate</CODE> - The replicate used for ratio calculations</DL>
</DD>
</DL>
<HR>

<A NAME="setPqs(double[])"><!-- --></A><H3>
setPqs</H3>
<PRE>
public void <B>setPqs</B>(double[]&nbsp;pqs)</PRE>
<DL>
<DD>Sets the probe set quality scores.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>pqs</CODE> - The probe set quality scores</DL>
</DD>
</DL>
<HR>

<A NAME="getPqs()"><!-- --></A><H3>
getPqs</H3>
<PRE>
public double[] <B>getPqs</B>()</PRE>
<DL>
<DD>Returns the probe set quality scores.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The probe set quality scores</DL>
</DD>
</DL>
<HR>

<A NAME="getAbsent()"><!-- --></A><H3>
getAbsent</H3>
<PRE>
public java.lang.String <B>getAbsent</B>()</PRE>
<DL>
<DD>Returns the call used for present/absent filtering.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The call used for present/absent filtering</DL>
</DD>
</DL>
<HR>

<A NAME="getMethod()"><!-- --></A><H3>
getMethod</H3>
<PRE>
public int <B>getMethod</B>()</PRE>
<DL>
<DD>Returns the number of the method used for ratio calculation.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>Number of the method used for ratio calculation</DL>
</DD>
</DL>
<HR>

<A NAME="getReplicate()"><!-- --></A><H3>
getReplicate</H3>
<PRE>
public int <B>getReplicate</B>()</PRE>
<DL>
<DD>Returns the number of the replicate used for calculations.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>Number of the replicate used for calculations</DL>
</DD>
</DL>
<HR>

<A NAME="getAbsentLabel()"><!-- --></A><H3>
getAbsentLabel</H3>
<PRE>
public java.lang.String <B>getAbsentLabel</B>()</PRE>
<DL>
<DD>Returns the name of the column with present/absent calls.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The name of the column with present/absent calls</DL>
</DD>
</DL>
<HR>

<A NAME="getGeneName()"><!-- --></A><H3>
getGeneName</H3>
<PRE>
public java.lang.String <B>getGeneName</B>()</PRE>
<DL>
<DD>Returns the name of the column with gene names.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The name of the column with gene names</DL>
</DD>
</DL>
<HR>

<A NAME="getSpotid()"><!-- --></A><H3>
getSpotid</H3>
<PRE>
public java.lang.String <B>getSpotid</B>()</PRE>
<DL>
<DD>Returns the name of the column with probeset ids.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The name of the column with probeset ids</DL>
</DD>
</DL>
<HR>

<A NAME="getMaximum()"><!-- --></A><H3>
getMaximum</H3>
<PRE>
public double <B>getMaximum</B>()</PRE>
<DL>
<DD>Returns the maximal expression value present in the data
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The maximal expression value</DL>
</DD>
</DL>
<HR>

<A NAME="getMinimum()"><!-- --></A><H3>
getMinimum</H3>
<PRE>
public double <B>getMinimum</B>()</PRE>
<DL>
<DD>Returns the minimal expression value present in the data
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The minimal expression value</DL>
</DD>
</DL>
<HR>

<A NAME="getFilename()"><!-- --></A><H3>
getFilename</H3>
<PRE>
public java.lang.String <B>getFilename</B>()</PRE>
<DL>
<DD>Returns the name of the file containing the data.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The name of the data file</DL>
</DD>
</DL>
<HR>

<A NAME="getSpot()"><!-- --></A><H3>
getSpot</H3>
<PRE>
public java.lang.String[] <B>getSpot</B>()</PRE>
<DL>
<DD>Returns the list of spots.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The list of spots</DL>
</DD>
</DL>
<HR>

<A NAME="getDat()"><!-- --></A><H3>
getDat</H3>
<PRE>
public double[] <B>getDat</B>()</PRE>
<DL>
<DD>Returns the evaluated normalized data.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The evaluated normalized data</DL>
</DD>
</DL>
<HR>

<A NAME="getAttributes()"><!-- --></A><H3>
getAttributes</H3>
<PRE>
public java.util.ArrayList&lt;<A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.String&gt;&gt; <B>getAttributes</B>()</PRE>
<DL>
<DD>Returns the attribute lists.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The list of attribute mappings</DL>
</DD>
</DL>
<HR>

<A NAME="getMap()"><!-- --></A><H3>
getMap</H3>
<PRE>
public <A HREF="../../halo/data/Mapping.html" title="class in halo.data">Mapping</A>&lt;java.lang.String,java.lang.Integer&gt; <B>getMap</B>()</PRE>
<DL>
<DD>Returns the map corresponding to the data.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The map for the data</DL>
</DD>
</DL>
<HR>

<A NAME="getAttrDescr()"><!-- --></A><H3>
getAttrDescr</H3>
<PRE>
public java.lang.String[] <B>getAttrDescr</B>()</PRE>
<DL>
<DD>Returns the description of the attributes.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The description of attributes</DL>
</DD>
</DL>
<HR>

<A NAME="getNewDescr()"><!-- --></A><H3>
getNewDescr</H3>
<PRE>
public java.lang.String[] <B>getNewDescr</B>()</PRE>
<DL>
<DD>Returns the description = column labels of newly transcribed RNA.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The description of newly transcribed RNA</DL>
</DD>
</DL>
<HR>

<A NAME="getNewRNA()"><!-- --></A><H3>
getNewRNA</H3>
<PRE>
public double[][] <B>getNewRNA</B>()</PRE>
<DL>
<DD>Returns all the values for newly transcribed RNA.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The measured values for newly transcribed RNA</DL>
</DD>
</DL>
<HR>

<A NAME="getCorrNewPre()"><!-- --></A><H3>
getCorrNewPre</H3>
<PRE>
public java.util.HashMap&lt;java.lang.String,java.lang.Double&gt; <B>getCorrNewPre</B>()</PRE>
<DL>
<DD>Returns the correction bias values for the ratio newly transcribed/pre-existing.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The correction bias values for the ratio newly transcribed/pre-existing</DL>
</DD>
</DL>
<HR>

<A NAME="getCorrNewTot()"><!-- --></A><H3>
getCorrNewTot</H3>
<PRE>
public java.util.HashMap&lt;java.lang.String,java.lang.Double&gt; <B>getCorrNewTot</B>()</PRE>
<DL>
<DD>Returns the correction bias values for the ratio newly transcribed/total.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The correction bias values for the ratio newly transcribed/total</DL>
</DD>
</DL>
<HR>

<A NAME="getCorrPreTot()"><!-- --></A><H3>
getCorrPreTot</H3>
<PRE>
public java.util.HashMap&lt;java.lang.String,java.lang.Double&gt; <B>getCorrPreTot</B>()</PRE>
<DL>
<DD>Returns the correction bias values for the ratio pre-existing/total.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The correction bias values for the ratio pre-existing/total</DL>
</DD>
</DL>
<HR>

<A NAME="getPreDescr()"><!-- --></A><H3>
getPreDescr</H3>
<PRE>
public java.lang.String[] <B>getPreDescr</B>()</PRE>
<DL>
<DD>Returns the description = column labels of pre-existing RNA.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The description of pre-existing RNA</DL>
</DD>
</DL>
<HR>

<A NAME="getPreRNA()"><!-- --></A><H3>
getPreRNA</H3>
<PRE>
public double[][] <B>getPreRNA</B>()</PRE>
<DL>
<DD>Returns all the values for pre-existing RNA.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The measured values for pre-existing RNA</DL>
</DD>
</DL>
<HR>

<A NAME="getTotalDescr()"><!-- --></A><H3>
getTotalDescr</H3>
<PRE>
public java.lang.String[] <B>getTotalDescr</B>()</PRE>
<DL>
<DD>Returns the description = column labeles of total RNA.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The description of total RNA</DL>
</DD>
</DL>
<HR>

<A NAME="getTotalRNA()"><!-- --></A><H3>
getTotalRNA</H3>
<PRE>
public double[][] <B>getTotalRNA</B>()</PRE>
<DL>
<DD>Returns all the values for total RNA.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The measured values for total RNA</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberExperiments()"><!-- --></A><H3>
getNumberExperiments</H3>
<PRE>
public int <B>getNumberExperiments</B>()</PRE>
<DL>
<DD>Returns the number of measured experiments
 (not changeable through filtering!).
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The number of measured experiments</DL>
</DD>
</DL>
<HR>

<A NAME="getSize()"><!-- --></A><H3>
getSize</H3>
<PRE>
public int <B>getSize</B>()</PRE>
<DL>
<DD>Returns the number of probesets present after filtering.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The number of (filtered) probesets</DL>
</DD>
</DL>
<HR>

<A NAME="getAvgLo()"><!-- --></A><H3>
getAvgLo</H3>
<PRE>
public double <B>getAvgLo</B>()</PRE>
<DL>
<DD>Returns the average log(e'/n').
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The average log(e'/n')</DL>
</DD>
</DL>
<HR>

<A NAME="getAvgUr()"><!-- --></A><H3>
getAvgUr</H3>
<PRE>
public double <B>getAvgUr</B>()</PRE>
<DL>
<DD>Returns the average number of Uracil.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The average number of Uracil</DL>
</DD>
</DL>
<HR>

<A NAME="getReplicatesTotal()"><!-- --></A><H3>
getReplicatesTotal</H3>
<PRE>
public int <B>getReplicatesTotal</B>()</PRE>
<DL>
<DD>Returns the total number of replicates.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The total number of replicates</DL>
</DD>
</DL>
<HR>

<A NAME="getPlotName()"><!-- --></A><H3>
getPlotName</H3>
<PRE>
public java.lang.String <B>getPlotName</B>()</PRE>
<DL>
<DD>Returns the name of the file into which evaluation
 results will be printed as a plottable table.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The name of the file containing the plot
 information for evaluation</DL>
</DD>
</DL>
<HR>

<A NAME="isUnequal()"><!-- --></A><H3>
isUnequal</H3>
<PRE>
public boolean <B>isUnequal</B>()</PRE>
<DL>
<DD>Returns TRUE if unequal number of RNA measurements has been loaded.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>TRUE if unequal number of RNA measurements has been loaded</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/Data.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../halo/data/Filter.html" title="class in halo.data"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?halo/data/Data.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Data.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
